File: iOS/AgoraEntScenarios/Common/API/CallAPI/CallAPI/Classes/CallMessageManager.swift
//
//  CallMessageManager.swift
//  CallAPI
//
//  Created by wushengtao on 2023/6/9.
//

import AgoraRtmKit

//let kReceiptsKey: String = "receipts"      //回执的消息id
//let kReceiptsRoomIdKey: String = "receiptsRoomId"     //回执到哪个房间，因为没有点对点，所以单点消息通过不同房间发送消息
let kMessageId: String = "messageId"     //发送的消息id

/// 回执的消息队列对象
//private class CallQueueInfo: NSObject {
//    var toUserId: String = ""
//    var messageId: Int = 0
//    var messageInfo: [String: Any]?
//    
//    var checkReceiptsFail: ((CallQueueInfo)->())?
//    
//    private var createDate: Date = Date()
//    private(set) var retryTimes: Int = 3
//    private var timer: Timer? {
//        didSet {
//            oldValue?.invalidate()
//        }
//    }
//    
//    deinit {
//        timer?.invalidate()
////        callMessagePrint("CallQueueInfo deinit \(messageId) cost: \(-Int(createDate.timeIntervalSinceNow * 1000)) ms")
//    }
//    
//    func checkReceipt() {
//        self.timer = Timer.scheduledTimer(withTimeInterval: 3, repeats: false) {[weak self] timer in
//            guard let self = self else {return}
//            self.retryTimes -= 1
////            callMessagePrint("receipt timeout retry \(self.retryTimes)")
//            self.checkReceiptsFail?(self)
//        }
//    }
//}

protocol CallMessageDelegate: NSObjectProtocol, AgoraRtmClientDelegate {
    func onConnectionFail()
    
    func debugInfo(message: String, logLevel: Int)
}

class CallMessageManager: NSObject {
    public weak var delegate: CallMessageDelegate?
    private var config: CallConfig!
    private var rtmClient: AgoraRtmClientKit!

    /// RTM是否已经登录
    private var isLoginedRtm: Bool = false
    
    /// 是否外部传入的rtm，如果是则不需要手动logout
    private var isExternalRtmClient: Bool = false
    
    private var prepareConfig: PrepareConfig?
    
    // 消息id
    private var messageId: Int = 0
    //待接收回执队列，保存没有接收到回执或者等待未超时的消息
//    private var receiptsQueue: [CallQueueInfo] = []
    
    deinit {
        #if DEBUG
        print("deinit-- CallMessageManager ")
        #endif
    }
    
    required init(config: CallConfig, delegate: CallMessageDelegate?) {
        super.init()
        self.config = config
        self.delegate = delegate
        if let rtmClient = config.rtmClient {
            //如果外部传入rtmclient，默认登陆成功
            self.isLoginedRtm = true
            self.isExternalRtmClient = true
            self.rtmClient = rtmClient
        } else {
            self.rtmClient = _createRtmClient(delegate: nil)
        }
        // disable retry message
        let _ = self.rtmClient.setParameters("{\"rtm.msg.tx_timeout\": 3000}")
        callMessagePrint("init-- CallMessageManager ")
    }
    
    private func _createRtmClient(delegate: AgoraRtmClientDelegate?) -> AgoraRtmClientKit {
        let rtmConfig = AgoraRtmClientConfig(appId: config!.appId, userId: "\(config!.userId)")
        if rtmConfig.userId.count == 0 {
            callMessagePrint("userId is empty", 2)
        }
        if rtmConfig.appId.count == 0 {
            callMessagePrint("appId is empty", 2)
        }

        var rtmClient: AgoraRtmClientKit? = nil
        do {
            rtmClient = try AgoraRtmClientKit(rtmConfig, delegate: nil)
        } catch {
            callMessagePrint("create rtm client fail: \(error.localizedDescription)", 2)
        }
        return rtmClient!
    }
}

//MARK: private method
extension CallMessageManager {
    //TODO: 超时时间是否可以自行设置(后端写死6s)
    //TODO: 保证时序否（保证）
    private func _sendMessage(userId: String, message: [String: Any], completion: ((NSError?)-> Void)?) {
        if userId.count == 0 {
            completion?(NSError(domain: "send message fail! roomId is empty", code: -1))
            return
        }
        let msgId = message[kMessageId] as? Int ?? 0
        let data = try? JSONSerialization.data(withJSONObject: message)
        let options = AgoraRtmPublishOptions()
        options.channelType = .user
        let date = Date()
        callMessagePrint("_sendMessage[\(msgId)] to '\(userId)', message: \(message)")
        rtmClient.publish(channelName: userId, data: data!, option: options) { [weak self] resp, err in
            guard let self = self else {return}
            if let err = err {
                let error = NSError(domain: err.reason, code: err.errorCode.rawValue)
                self.callMessagePrint("_sendMessage[\(msgId)] fail: \(error) cost: \(date.getCostMilliseconds()) ms", 1)

                completion?(error)
                return
            }
            self.callMessagePrint("_sendMessage[\(msgId)] publish cost \(date.getCostMilliseconds()) ms")
            completion?(nil)
        }
    }
    
    private func loginRtm(rtmClient: AgoraRtmClientKit, token: String, completion: @escaping (AgoraRtmErrorInfo?)->()) {
        if isLoginedRtm {
            completion(nil)
            return
        }
        rtmClient.logout()
        callMessagePrint("will login")
        rtmClient.login(token) {[weak self] resp, error in
            guard let self = self else {return}
            self.callMessagePrint("login completion: \(error?.errorCode.rawValue ?? 0)")
            self.isLoginedRtm = error == nil ? true : false
            completion(error)
        }
    }
}

//MARK: public method
extension CallMessageManager {
    func deinitialize() {
        rtmClient.removeDelegate(self)
        if isExternalRtmClient == false {
            rtmClient.logout()
            rtmClient.destroy()
        }
    }
    
    /// 根据配置初始化RTM
    /// - Parameters:
    ///   - prepareConfig: <#prepareConfig description#>
    ///   - tokenConfig: <#tokenConfig description#>
    ///   - completion: <#completion description#>
    func initialize(prepareConfig: PrepareConfig, completion: ((NSError?) -> ())?) {
        callMessagePrint("initialize")
        self.prepareConfig = prepareConfig
        rtmClient.addDelegate(self)
        let rtmToken = prepareConfig.rtmToken
        if rtmToken.isEmpty, isExternalRtmClient == false {
            let reason = "RTM Token is Empty"
            completion?(NSError(domain: reason, code: -1))
            return
        }
        
        guard let rtmClient = self.rtmClient else {
            let reason = "rtmClient is nil, please invoke 'initialize' to setup config"
            completion?(NSError(domain: reason, code: -1))
            return
        }
        
        if !isLoginedRtm {
            loginRtm(rtmClient: rtmClient, token: rtmToken) {/*[weak self]*/ err in
                if let err = err, err.errorCode != .ok {
                    completion?(NSError(domain: err.reason, code: err.errorCode.rawValue))
                    return
                }

                completion?(nil)
            }
        } else {
            completion?(nil)
        }
    }
    
    /// 更新RTM token
    /// - Parameter tokenConfig: CallTokenConfig
    public func renewToken(rtcToken: String, rtmToken: String) {
        guard isLoginedRtm else {
            if let prepareConfig = prepareConfig {
                //没有登陆成功，但是需要自动登陆，可能是初始token问题，这里重新initialize
                callMessagePrint("renewToken need to reinit")
                self.rtmClient.logout()
                initialize(prepareConfig: prepareConfig) { err in
                }
            }
            return
        }
        self.prepareConfig?.rtcToken = rtcToken
        self.prepareConfig?.rtmToken = rtmToken
        rtmClient?.renewToken(rtmToken, completion: {[weak self] resp, err in
            self?.callMessagePrint("rtm renewToken: \(err?.errorCode.rawValue ?? 0)")
        })
    }
    
    /// 发送频道消息
    /// - Parameters:
    ///   - userId: 往哪个用户发送消息
    ///   - message: 发送的消息字典
    ///   - completion: <#completion description#>
    public func sendMessage(userId: String,
                            message: [String: Any],
                            completion: ((NSError?)-> Void)?) {
        guard userId.count > 0, userId != "0" else {
            let errorStr = "sendMessage fail, invalid userId[\(userId)]"
            callMessagePrint(errorStr)
            completion?(NSError(domain: errorStr, code: -1))
            return
        }
        
        messageId += 1
        messageId %= Int.max
        var message = message
        message[kMessageId] = messageId
//        message[kReceiptsRoomIdKey] = fromUserId
//        assert(fromUserId.count > 0, "kReceiptsRoomIdKey is empty")
        _sendMessage(userId: userId, message: message, completion: completion)
    }
}

//MARK: AgoraRtmClientDelegate
extension CallMessageManager: AgoraRtmClientDelegate {
    func rtmKit(_ kit: AgoraRtmClientKit, channel channelName: String, connectionChangedToState state: AgoraRtmClientConnectionState, reason: AgoraRtmClientConnectionChangeReason) {
        callMessagePrint("rtm connectionChangedToState: \(state.rawValue) reason: \(reason.rawValue)")
        if reason == .changedTokenExpired {
            self.delegate?.rtmKit?(kit, tokenPrivilegeWillExpire: nil)
        } else if reason == .changedChangedLost {
            self.delegate?.onConnectionFail()
        }
    }
    
    func rtmKit(_ rtmKit: AgoraRtmClientKit, tokenPrivilegeWillExpire channel: String?) {
        callMessagePrint("rtm onTokenPrivilegeWillExpire[\(channel ?? "nil")]")
        self.delegate?.rtmKit?(rtmKit, tokenPrivilegeWillExpire: channel)
    }
    
    //收到RTM消息
    public func rtmKit(_ rtmKit: AgoraRtmClientKit, didReceiveMessageEvent event: AgoraRtmMessageEvent) {
//        let message = event.message
//        if let data = message.rawData,
//           let dic = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
//            if let messageId = dic[kMessageId] as? Int,
//               let receiptsRoomId = dic[kReceiptsRoomIdKey] as? String {
//                _sendReceipts(roomId: receiptsRoomId, messageId: messageId)
//            } else if let receiptsId = dic[kReceiptsKey] as? Int {
//                receiptsQueue = receiptsQueue.filter({$0.messageId != receiptsId})
//                callMessagePrint("recv receipts: \(receiptsId) msgIds\(self.receiptsQueue.map({ $0.messageId}))")
//            }
//            
//            callMessagePrint("on event message: \(String(data: data, encoding: .utf8) ?? "")")
//        } else {
//            callMessagePrint("on event message parse fail, \(message.rawData?.count ?? 0) )", 1)
//        }
        
        self.delegate?.rtmKit?(rtmKit, didReceiveMessageEvent: event)
    }
}

extension CallMessageManager {
    private func callMessagePrint(_ message: String, _ logLevel: Int = 0) {
        let tag = "[MessageManager][\(String.init(format: "%p", self))][\(String.init(format: "%p", rtmClient))]"
        delegate?.debugInfo(message: "\(tag)\(message)", logLevel: logLevel)
        #if DEBUG
        if let _ = delegate {return}
//        print("\(formatter.string(from: Date()))[CallApi]\(tag)\(message)")
        #endif
    }
}

File: Android/scenes/showTo1v1/src/main/java/io/agora/scene/showTo1v1/callAPI/CallMessageManager.kt
package io.agora.scene.showTo1v1.callAPI

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.gson.Gson
import io.agora.rtm.*
import org.json.JSONObject

/// 回执的消息队列对象
private class CallQueueInfo {

    val TAG = "CALL_QUEUE_LOG"
    var messageId: Int = 0
    var messageInfo: Map<String, Any>? = null
    var callback: ((AGError?) -> Unit)? = null

    var checkReceiptsFail: ((CallQueueInfo) -> Unit)? = null

    private val createDate = System.currentTimeMillis()
    var retryTimes: Int = 3
    private val mHandler = Handler(Looper.getMainLooper())
    private var retryRunnable: Runnable? = null
        set(value) {
            field?.let { mHandler.removeCallbacks(it) }
            field = value
            field?.let { mHandler.postDelayed(it, 3000) }
        }

    fun checkReceipt() {
        retryRunnable = Runnable {
            retryTimes -= 1
            Log.d(TAG, "receipt timeout retry $retryTimes , message id: $messageId")
            checkReceiptsFail?.invoke(this@CallQueueInfo)
        }
    }

    init {
        Log.d(TAG, "CallQueueInfo init $messageId")
    }

    fun finish() {
        retryRunnable?.let { mHandler.removeCallbacks(it) }
        Log.d(TAG, "CallQueueInfo deinit $messageId cost: $createDate ms")
    }
}

interface CallMessageListener: RtmEventListener {
    /** 回执没有收到*/
    fun onMissReceipts(message: Map<String, Any>)
    fun onConnectionFail()
    fun debugInfo(message: String)
    fun debugWarning(message: String)
}

class CallMessageManager(
    private val config: CallConfig,
    private val rtmClient: RtmClient,
    private var listener: CallMessageListener? = null,
): RtmEventListener {

    private val TAG = "CALL_MSG_MANAGER"

    /** 回执的消息id */
    private val kReceiptsKey = "receipts"

    /** 回执到哪个房间，因为没有点对点，所以单点消息通过不同房间发送消息 */
    private val kReceiptsRoomIdKey = "receiptsRoomId"

    /** 发送的消息id */
    private val kMessageId = "messageId"
    // RTM是否已经登录
    private var isLoginedRTM = false
    // RTM 是否已经订阅频道
    var isSubscribedRTM = false
        private set

    private var prepareConfig: PrepareConfig? = null
    private var tokenConfig: CallTokenConfig? = null
    // 消息id
    private var messageId: Int = 0
    // 待接收回执队列，保存没有接收到回执或者等待未超时的消息
    private var receiptsQueue: MutableList<CallQueueInfo>? = mutableListOf()

    private val mHandler = Handler(Looper.getMainLooper())
    init {
        if (rtmClient == config.rtmClient) {
            //如果外部传入rtmclient，默认登陆成功
            isLoginedRTM = true
        }
        callMessagePrint("init-- CallMessageManager ")
    }

    fun rtmDeinitialize() {
        rtmClient.removeEventListener(this)
        callMessagePrint("unsubscribe[${config.userId}]")
        receiptsQueue?.forEach { it.finish() }
        receiptsQueue = null
        rtmClient.unsubscribe(config.userId.toString(), object : ResultCallback<Void> {
            override fun onSuccess(responseInfo: Void?) {}
            override fun onFailure(errorInfo: ErrorInfo?) {}
        })
    }

    fun rtmInitialize(prepareConfig: PrepareConfig, tokenConfig: CallTokenConfig?, completion: (AGError?) -> Unit) {
        callMessagePrint("_rtmInitialize")
        this.prepareConfig = prepareConfig
        this.tokenConfig = tokenConfig
        rtmClient.addEventListener(this)
        val rtmToken = tokenConfig?.rtmToken ?: run {
            val reason = "RTM Token is Empty"
            completion(AGError(reason, -1))
            return
        }
        val rtmClient = this.rtmClient
        if (prepareConfig.autoLoginRTM && !isLoginedRTM) {
            loginRTM(rtmClient, rtmToken) { err ->
                if (err != null) {
                    val errorCode = RtmConstants.RtmErrorCode.getValue(err.errorCode)
                    completion.invoke(AGError(err.errorReason, errorCode))
                }
                rtmInitialize(prepareConfig, tokenConfig, completion)
            }
        } else if (isLoginedRTM && prepareConfig.autoSubscribeRTM) {
            _subscribeRTM(config.userId.toString()) { error ->
                if (error != null) {
                    completion.invoke(error)
                } else {
                    isSubscribedRTM = true
                    completion.invoke(null)
                }
            }
        } else {
            completion.invoke(null)
        }
    }

    /// 更新RTM token
    /// - Parameter rtmToken: <#rtmToken description#>
    fun renewToken(tokenConfig: CallTokenConfig) {
        if (!isLoginedRTM) {
            val prepareConfig = prepareConfig
            if (prepareConfig != null && prepareConfig.autoLoginRTM) {
                //没有登陆成功，但是需要自动登陆，可能是初始token问题，这里重新initialize
                callMessagePrint("renewToken need to reinit")
                rtmClient.logout(object : ResultCallback<Void> {
                    override fun onSuccess(responseInfo: Void?) {}
                    override fun onFailure(errorInfo: ErrorInfo?) {}
                })
                rtmInitialize(prepareConfig, tokenConfig) { _ ->
                }
            }
            return
        }
        this.tokenConfig = tokenConfig
        rtmClient.renewToken(tokenConfig.rtmToken, object : ResultCallback<Void> {
            override fun onSuccess(responseInfo: Void?) {
                callMessagePrint("rtm renewToken")
            }
            override fun onFailure(errorInfo: ErrorInfo?) {
            }
        })
    }

    /** 发送频道消息
     * @param userId: 往哪个用户发送消息
     * @param fromRoomId: 哪个频道发送的，用来给对端发送回执
     * @param message: 发送的消息字典
     * @param completion: <#completion description#>
     */
    fun sendMessage(userId: String, fromUserId: String, message: Map<String, Any>, completion: ((AGError?)-> Unit)?) {
        messageId += 1
        messageId %= Int.MAX_VALUE
        val map = message.toMutableMap()
        val msgId = messageId
        map[kMessageId] = msgId
        map[kReceiptsRoomIdKey] = fromUserId
        require(fromUserId.isNotEmpty()) { "kReceiptsRoomIdKey is empty" }
        _sendMessage(userId, map, completion)
    }

    /** 根据策略订阅频道消息
     *  @param userId: 频道号
     *  @param completion: <#completion description#>
     */
    private fun _subscribeRTM(userId: String, completion: ((AGError?) -> Unit)?) {
        val options = SubscribeOptions()
        options.withMessage = true
        options.withMetadata = false
        options.withPresence = false
        callMessagePrint("will subscribe[$userId] message: ${options.withMessage} presence: ${options.withPresence}")
        rtmClient.unsubscribe(userId, object: ResultCallback<Void> {
            override fun onSuccess(responseInfo: Void?) {}
            override fun onFailure(errorInfo: ErrorInfo?) {}
        })
        rtmClient.subscribe(userId, options, object: ResultCallback<Void> {
            override fun onSuccess(responseInfo: Void?) {
                callMessagePrint("subscribe[$userId] finished = success")
                runOnUiThread { completion?.invoke(null) }
            }
            override fun onFailure(errorInfo: ErrorInfo?) {
                val msg = errorInfo?.errorReason ?: "error"
                val errorCode = RtmConstants.RtmErrorCode.getValue(errorInfo?.errorCode)
                runOnUiThread { completion?.invoke(AGError(msg, errorCode)) }
            }
        })
    }

    /** 发送回执消息
     * @param roomId: 回执消息发往的频道
     * @param messageId: 回执的消息id
     * @param completion: <#completion description#>
     */
    private fun _sendReceipts(roomId: String, messageId: Int, completion: ((AGError?)-> Unit)? = null) {
        val message = mutableMapOf<String, Any>()
        message[kReceiptsKey] = messageId
        callMessagePrint("_sendReceipts to $roomId, message: $message")
        val json = Gson().toJson(message)
        val options = PublishOptions()
        rtmClient.publish(roomId, json.toByteArray(), options, object : ResultCallback<Void> {
            override fun onSuccess(responseInfo: Void?) {
                callMessagePrint("_sendReceipts cost ${System.currentTimeMillis()} ms")
                runOnUiThread { completion?.invoke(null) }
            }
            override fun onFailure(errorInfo: ErrorInfo?) {
                val msg = errorInfo?.errorReason ?: "error"
                val errorCode = RtmConstants.RtmErrorCode.getValue(errorInfo?.errorCode)
                runOnUiThread { completion?.invoke(AGError(msg, errorCode)) }
            }
        })
    }

    private fun _sendMessage(userId: String, message: Map<String, Any>, completion:((AGError?)->Unit)?) {
        if (userId.isEmpty()) {
            completion?.invoke(AGError("send message fail! roomId is empty", -1))
            return
        }
        callMessagePrint("_sendMessage to '$userId', message: $message")
        val msgId = message[kMessageId] as? Int ?: 0
        val json = Gson().toJson(message)
        val options = PublishOptions()
        val startTime = System.currentTimeMillis()
        rtmClient.publish(userId, json.toByteArray(), options, object : ResultCallback<Void> {
            override fun onSuccess(p0: Void?) {
                callMessagePrint("publish cost ${System.currentTimeMillis() - startTime} ms")
                runOnUiThread { completion?.invoke(null) }
                //发送成功检查回执，没收到则重试
                receiptsQueue?.let { queue ->
                    queue.firstOrNull { it.messageId == msgId }?.let {
                        it.checkReceipt()
                        return
                    }
                    val receiptInfo = CallQueueInfo()
                    receiptInfo.messageId = msgId
                    receiptInfo.messageInfo = message
                    receiptInfo.callback = completion
                    receiptInfo.checkReceiptsFail = { info ->
                        if (info.retryTimes > 0) {
                            _sendMessage(userId, message, completion = completion)
                        } else {
                            val messageInfo = info.messageInfo ?: emptyMap()
                            queue.filter {it.messageId == msgId}.forEach { it.finish() }
                            queue.removeAll { it.messageId == msgId }
                            listener?.onMissReceipts(messageInfo)
                        }
                    }
                    queue.add(receiptInfo)
                    receiptInfo.checkReceipt()
                }
            }

            override fun onFailure(errorInfo: ErrorInfo?) {
                val msg = errorInfo?.errorReason ?: "error"
                val code = errorInfo?.errorCode?.ordinal ?: -1
                callWarningPrint("_sendMessage: fail: $msg")
                runOnUiThread { completion?.invoke(AGError(msg, code)) }
            }
        })
    }
    private fun loginRTM(rtmClient: RtmClient, token: String, completion: (ErrorInfo?) -> Unit) {
        if (isLoginedRTM) {
            completion(null)
            return
        }
        callMessagePrint("will login")
        rtmClient.logout(object : ResultCallback<Void?> {
            override fun onSuccess(responseInfo: Void?) {}
            override fun onFailure(errorInfo: ErrorInfo?) {}
        })
        rtmClient.login(token, object : ResultCallback<Void?> {
            override fun onSuccess(p0: Void?) {
                callMessagePrint("login completion")
                isLoginedRTM = true
                runOnUiThread { completion(null) }
            }

            override fun onFailure(p0: ErrorInfo?) {
                callMessagePrint("login completion: ${p0?.errorCode}")
                isLoginedRTM = false
                runOnUiThread { completion(p0) }
            }
        })
    }
    //MARK: AgoraRtmClientDelegate
    override fun onConnectionStateChanged(
        channelName: String?,
        state: RtmConstants.RtmConnectionState?,
        reason: RtmConstants.RtmConnectionChangeReason?
    ) {
        callMessagePrint("rtm connectionStateChanged: $state reason: $reason")
        runOnUiThread {
            if (reason == RtmConstants.RtmConnectionChangeReason.TOKEN_EXPIRED) {
                listener?.onTokenPrivilegeWillExpire(channelName)
            } else if (reason == RtmConstants.RtmConnectionChangeReason.LOST) {
                listener?.onConnectionFail()
            }
        }
    }
    override fun onTokenPrivilegeWillExpire(channelName: String?) {
        callMessagePrint("rtm onTokenPrivilegeWillExpire[${channelName ?: "nil"}]")
        runOnUiThread {
            listener?.onTokenPrivilegeWillExpire(channelName)
        }
    }

    override fun onMessageEvent(event: MessageEvent?) {
        val message = event?.message?.data as? ByteArray ?: return
        val jsonString = String(message, Charsets.UTF_8)
        callMessagePrint("on event message: $jsonString")
        val map = jsonStringToMap(jsonString)
        val messageId = map[kMessageId] as? Int
        val receiptsRoomId = map[kReceiptsRoomIdKey] as? String
        val receiptsId = map[kReceiptsKey] as? Int
        if (receiptsRoomId != null && messageId != null) {
            _sendReceipts(receiptsRoomId, messageId)
        } else if (receiptsId != null) {
            callMessagePrint("recv receipts $receiptsId")
            receiptsQueue?.let { queue ->
                queue.filter {it.messageId == receiptsId}.forEach { it.finish() }
                queue.removeAll { it.messageId == receiptsId }
            }
        } else {
            callWarningPrint("on event message parse fail, ${event.message.type} ${event.message.data}")
        }
        runOnUiThread {
            listener?.onMessageEvent(event)
        }
    }

    override fun onPresenceEvent(event: PresenceEvent?) {
        runOnUiThread {
            listener?.onPresenceEvent(event)
        }
    }

    override fun onTopicEvent(event: TopicEvent?) {}
    override fun onLockEvent(event: LockEvent?) {}
    override fun onStorageEvent(event: StorageEvent?) {}
    private fun jsonStringToMap(jsonString: String): Map<String, Any> {
        val json = JSONObject(jsonString)
        val map = mutableMapOf<String, Any>()
        val keys = json.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            map[key] = json.get(key)
        }
        return map
    }

    private fun callMessagePrint(message: String) {
        val tag = "[MessageManager]"
        listener?.debugInfo("$tag$message)")
        if (listener == null) {
            Log.d(TAG, "[CallApi]$tag $message)")
        }
    }

    private fun callWarningPrint(message: String) {
        val tag = "[MessageManager]"
        listener?.debugWarning("$tag$message")
        if (listener == null) {
            Log.d(TAG, "[CallApi][Warning]$tag$message")
        }
    }
    private fun runOnUiThread(runnable: Runnable) {
        if (Thread.currentThread() == Looper.getMainLooper().thread) {
            runnable.run()
        } else {
            mHandler.post(runnable)
        }
    }
}


